### DISCUSSION TITLE
Is it possible to emulate functionality of vim but desktop-wide?

### DISCUSSION THREAD

**Discussion by laicasaane** [houmain/keymapper#109] [Q&A] [answered]
This question has just come up at my workplace and after a bit of researching around GitHub I found keymapper. But from the document and examples, I'm still unsure if keymapper can emulate vim.

For example, I want to navigate my desktop, applications and web brower by simple keystrokes (without <kbd>Ctrl</kbd>, <kbd>Shift</kbd>, etc.). I can also switch to input mode for text typing.

**Comment by ristomatti** [comment]
It surely depends what level of emulation you are hoping to create. Some sort of normal/input mode switching should definitely be possible with the virtual modifier feature.

**Reply by laicasaane** [reply]
Thank, I'll give keymapper a try.

**Reply by ristomatti** [reply]
Out of curiosity, I created a proof of context config with a few "vimish" features for you to get an idea of what you could do. It's intended to be used in a traditional text editor on Linux/Windows.

The commented out `log`, `context` and `ContextActive` would require creatinga  custom build from the `next` branch to work. I left them there as they were useful for debugging the config. Related to https://github.com/houmain/keymapper/issues/91#issuecomment-1946504547

`CapsLock` toggles the Vim emulation mode on/off. 

```perl
# define some aliases
Alt   = AltLeft

# immediately forward common modifier keys (also used as mouse modifiers)
Shift   >> Shift
Control >> Control
Alt     >> Alt

# log = $(echo "$0" >> ~/tmp/keymapper.log)
# context = 50ms log["entered $0"] ^ log["left $0"]

VimMode    = Virtual0
NormalMode = Virtual1
InsertMode = Virtual2
VisualMode = Virtual3

[modifier="VimMode"]
  # ContextActive >> context["VimMode"]

[modifier="VimMode InsertMode"]
  # ContextActive >> context["InsertMode"]
  Escape      >> InsertMode NormalMode
  Control{C}  >> InsertMode NormalMode

  CapsLock >> InsertMode VimMode

[modifier="VimMode VisualMode"]
  # ContextActive >> context["VisualMode"]
  H           >> Shift{ArrowLeft}
  J           >> Shift{ArrowDown}
  K           >> Shift{ArrowUp}
  L           >> Shift{ArrowRight}
  Shift{6}    >> Shift{Home}
  Shift{4}    >> Shift{End}

  X           >> Control{X} VisualMode NormalMode
  Y           >> Control{C} ArrowLeft VisualMode NormalMode
  P           >> Control{V} VisualMode NormalMode

  Escape      >> VisualMode NormalMode
  Control{C}  >> VisualMode NormalMode

  CapsLock    >> VisualMode VimMode

  Any         >> Shift{Any}

[modifier="VimMode NormalMode"]
  # ContextActive >> context["NormalMode"]
  I           >> NormalMode InsertMode
  V           >> NormalMode VisualMode
  H           >> ArrowLeft
  J           >> ArrowDown
  K           >> ArrowUp
  L           >> ArrowRight
  X           >> Delete
  Y           >> Control{C}
  P           >> Control{V}
  U           >> Control{Z}
  R           >> Control{Y}
  Shift{6}    >> Home
  Shift{4}    >> End

  CapsLock    >> NormalMode VimMode

[default]
  # ContextActive >> context["Default"]
  CapsLock    >> VimMode NormalMode
```

**Reply by laicasaane** [reply]
Thanks. However, I'm not sure if you follow me, so I have to ask for confirmation. My idea is using keymapper to enable desktop navigation via keyboard only (just like vim text editor). For example, navigate to my web browser, scrolling a web page, open new tab, switch to input mode to enter a new address, navigate to file explorer, open a folder or file, etc. 🤔

**Reply by ristomatti** [reply]
Which OS?

**Reply by ristomatti** [reply]
Are you aware of the Vimium Chrome/Firefox extension BTW https://vimium.github.io/? It solves your browser related needs quite nicely. Although it does not work on a new tab page which kind of cuts the flow at some points. I've set up some Keymapper bindings to do various stuff in Chrome.

I've tried to add all the aliases used in these keymaps from the top section of my config.

```perl
AltGr     = AltRight
EitherAlt = AltLeft | AltGr

# Alias to prevent modifiers used in keymappings affect the mapped key
_Control  = !ControlRight !MetaLeft !MetaRight !AltLeft !AltGr ControlLeft
_Alt      = !ControlLeft !ControlRight !MetaLeft !MetaRight !AltGr AltLeft

# Directional navigation
KeyUp           = I
KeyLeft         = J
KeyDown         = K
KeyRight        = L
Direction       = KeyUp | KeyLeft | KeyDown | KeyRight

# Navigation layer activated when Escape is pressed
NavUp           = Numpad8
NavLeft         = Numpad4
NavDown         = Numpad2
NavRight        = Numpad6

Digit     = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9

# This could be any key but Escape is convenient for me as I've got a split
# onto which I've put Escape between the space spacebars
Leader    = Escape

[class=/Chrome|Chromium/i]
  Leader{Digit}         >> AltLeft{Digit}
  Leader{N}             >> new_window
  Leader{T}             >> new_tab
  Leader{W}             >> close_tab
  Leader{G}             >> goto_location
  Leader{Z}             >> suspend_tab
  Leader{B}             >> navigate_back
  Leader{F}             >> navigate_forward
  Leader{R}             >> reload_page
  Leader{KeyLeft}       >> prev_tab
  Leader{KeyRight}      >> next_tab
  Leader{KeyUp}         >> shift_tab_left
  Leader{KeyDown}       >> shift_tab_right

  # You could have K and J here if you'd create a modal setup,
  # but Vimium works for this just great 
  NavDown               >> scroll_up
  NavLeft               >> scroll_down

  EitherAlt{N}          >> new_window
  EitherAlt{T}          >> new_tab
  EitherAlt{W}          >> close_tab

  AltGr{R}              >> reload_page
  AltGr{P}              >> pin_tab
  AltGr{Z}              >> suspend_tab

  Alt{KeyLeft}          >> prev_tab
  Alt{KeyRight}         >> next_tab
  Alt{KeyUp}            >> scroll_up
  Alt{KeyDown}          >> scroll_down

  goto_location         >> _Control{L}
  shift_tab_left        >> !AltGr Control{Shift{PageUp}}
  shift_tab_right       >> !AltGr Control{Shift{PageDown}}
  pin_tab               >> !AltGr AltLeft{Shift{X}}
  reload_page           >> _Control{R}
  scroll_up             >> ArrowUp ArrowUp ArrowUp ArrowUp ArrowUp
  scroll_down           >> ArrowDown ArrowDown ArrowDown ArrowDown ArrowDown

  new_window            >> _Control{N}
  new_tab               >> _Control{T}
  close_tab             >> _Control{W}
  prev_tab              >> _Control{PageUp}
  next_tab              >> _Control{PageDown}
  navigate_back         >> _Alt{ArrowLeft}
  navigate_forward      >> _Alt{ArrowRight}

```

**Reply by ristomatti** [reply]
Personally I use Linux and i3wm and use mouse very little but that kind of comes already with i3wm. In case you're using Windows, you might [AutoHotkey](https://www.autohotkey.com/) useful for working out opening specific apps etc. It's what I used some 10y ago when I still had Windows.

You can take a look at a slightly dated version of my full config here in case you're interested https://gist.github.com/ristomatti/7295775781ad697bd274bad858427281. 

:warning: The config has quite a bit going on and might just be confusing.

I've been experimenting with a lot things. It will likely end up much simpler based on what keymaps actually stick to muscle memory and which not.

**Reply by ristomatti** [reply]
Oh, another thing to note about the `Leader` mappings on the Chrome example. The syntax used actually means `Leader+X` but I actually press `Leader` then `X` in sequence. It works for me but it might be due to having a mechanical keyboard or the fact my keyboard runs QMK firmware. If the same doesn't work for you, try modifying them to something like this:

```perl
Leader    = Escape !200ms

[class=/Chrome|Chromium/i]
  Leader Digit         >> AltLeft{Digit}
  Leader N             >> new_window
  Leader T             >> new_tab

  # ...other mappings

  # Avoid sending mistyped characters
  Leader Any           >> Leader
  Leader               >> Leader
```

In this way it is more correct but for some reason `Leader{X}` works more reliably for me.

**Reply by ristomatti** [reply]
> Personally I use Linux and i3wm and use mouse very little but that kind of comes already with i3wm. In case you're using Windows, you might [AutoHotkey](https://www.autohotkey.com/) useful for working out opening specific apps etc. It's what I used some 10y ago when I still had Windows.

If you're on Windows/MacOS, there's tiling window managers for them as well. Tiling window managers to be focused on keyboard navigation. Some popular options seem to be [komorebi](https://github.com/LGUG2Z/komorebi) for Windows, and [yabai](https://github.com/koekeishiya/yabai) for MacOS. I haven't tried either but I likely would if I had to use Windows/Mac.

**Reply by laicasaane** [reply]
I'm using Windows. Tiling window manager is not what I want though. I'm aware of Vimium extension but it's limited to the browser only. I know AutoHotkey and had used it a bit in the past, but it may cause problem with anti-cheat programs (btw, not an important thing though).

**Reply by ristomatti** [reply, marked as answer]
It's likely too well known since it's old as time. I'm not a gamer myself but my intention was to suggest AutoHotkey for handling window management, like focusing a specific window from a shortcut etc. since Keymapper is focused on remapping keys to other keys. It can run scripts/apps as well but to e.g. bring an app to focus, you'd need to have some utility to do that.

From the last section of my config you'll find some mappings using `i3-msg` to give instructions to i3 window manager. You need to find an equivalent for Windows to be able to do what you describe I believe. Unless of course it's enough to fill the use case by mapping other keys to `Win+1-9` that launch/open the apps you've pinned on the start bar. 🙂

Going back to AutoHotkey, I know it's syntax has changed with the new 2.0 version but to give some idea what it can do with window management, you can check a config I wrote some ~15y ago (with the older syntax) https://gist.github.com/ristomatti/f075020025cf870aaf8f73a930996f14.

What it does (or did) is allow moving/resizing windows using `Win+left/right/middle` mouse button based on the part of the window you click. Left click moves the window based on the clicked section if the window in a 3x3 grid. Right click extends the the window to the clicked direction or if clicked on the center,it makes the window 50% width/height of the screen (to allow making a quick split, 2x2, 1x2 etc layout). Middle click works in two steps. First click selects a window and the second one the place to move and fit the Window. E.g. when clicked on an empty space, it resizes the window to fit the available space. When the second click is a window, it resizes the other window to the same size. So it was a poor mans tiling WM back in the time when Windows didn't have this feature.

Another thing it does is directional window focus using `Win+up/down/left/right` arrow keys. I do not know if Windows has this feature even to this day as alt tab still seems to be the cool way to do it(?) 😄

**Reply by laicasaane** [reply]
I guess I'd have to experiment a bit. Thanks!

**Reply by rp1231** [reply]
There is this software that aims to do just that:
https://github.com/pit-ray/win-vind
I had used it quite a while back when it was a bit slow.
But it seems to have continued development, so maybe it's faster now?

**Comment by codiggermkt** [comment]
Yes, it is possible to emulate Vim-like functionality desktop-wide by using various tools and applications that provide similar keybindings and behaviors to Vim. Here are some ways to achieve Vim-like functionality outside of the Vim editor:

1、Use Vim Keybindings in Other Applications:
2、Tiling Window Managers
3、Custom Keyboard Shortcuts
4、Vim Emulation Software

**Comment by Zireael07** [comment]
Since this was posted, someone shared their keymapper vim inspired config: https://www.nik.tw/post/achieving-the-holy-grail-vim-bindings-everywhere/
